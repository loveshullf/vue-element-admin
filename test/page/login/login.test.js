import { mount, createLocalVue } from "vue-test-utils";

const localVue = createLocalVue();
import ElementUI from "element-ui";
import Vuex from "vuex";
localVue.use(Vuex);
localVue.use(ElementUI);

//模拟数据以及axios
// import "@/mocks";
// jest.mock('@/api/login'); 
import {login} from '@/api/login';
localVue.prototype.$http = login;

//store moudel
jest.mock("@/store/app/modules/user.js");
import user from "@/store/app/modules/user.js";

let wrapper;
import Login from "@/page/login";
beforeEach(() => {
  let store = new Vuex.Store({
    modules:{user}
  });
  wrapper = mount(Login, { store, localVue });
 //每次测试前确保我们的测试实例都是是干净完整的。返回一个wrapper对象
});

test("界面应该有两个input和一个button", () => {
  const inputs = wrapper.findAll(".el-input"); // 通过findAll来查找dom或者vue实例
  const loginButton = wrapper.find(".el-button"); // 通过find查找元素
  expect(inputs.length).toBe(2); // 应该有两个输入框
  expect(loginButton).toBeTruthy(); // 应该有一个登录按钮。 只要断言条件不为空或这false，toBeTruthy就能通过。
});

test("登录按钮,触发handleLogin方法", () => {
  const stub = jest.fn(); // 伪造一个jest的mock funciton
  wrapper.setMethods({ handleLogin: stub });
  wrapper.find(".el-button").trigger("click"); // 对button触发一个click事件
  expect(stub).toBeCalled(); // 查看loginToDo是否被调用
});

test("验证输入格式，错误格式输入返回FALSE", () => {
  wrapper.setData({
    loginForm: {
      username: "admin",
      password: "1234"
    }
  }); //模拟用户输入数据
  const stub=jest.fn((valid)=>{ expect(valid).toBe(false)})
  wrapper.vm.$refs.loginForm.validate(stub);
});

test("vuex状态测试，登录按钮按下验证通过之后，Action:LoginByUsername被调用一次", () => {
  expect.assertions(1);
  wrapper.setData({
    loginForm: {
      username: "admin",
      password: "123456"
    }
  }); //模拟用户输入数据
  let spy=jest.spyOn(user.actions,'LoginByUsername');
  wrapper.find(".el-button").trigger("click");
  expect(spy).toHaveBeenCalledTimes(1);
});
//快照
test("页面快照", () => {
  expect(wrapper.element).toMatchSnapshot(); // 调用toMatchSnapshot来比对快照
});
