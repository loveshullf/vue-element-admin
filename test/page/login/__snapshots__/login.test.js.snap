// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`页面快照 1`] = `
<div
  class="login-container"
>
  <form
    autocomplete="on"
    class="el-form card-box login-form el-form--label-left"
  >
    <h3
      class="title"
    >
      系统登录
    </h3>
     
    <div
      class="el-form-item is-required"
    >
      <!---->
      <div
        class="el-form-item__content"
      >
        <span
          class="svg-container svg-container_login"
        >
          <svg-icon
            svg-name="user"
          />
        </span>
         
        <div
          class="el-input"
        >
          <!---->
          <input
            autocomplete="on"
            class="el-input__inner"
            name="username"
            placeholder="用户名/邮箱"
            rows="2"
            type="text"
            validateevent="true"
          />
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
     
    <div
      class="el-form-item is-required"
    >
      <!---->
      <div
        class="el-form-item__content"
      >
        <span
          class="svg-container"
        >
          <svg-icon
            svg-name="password"
          />
        </span>
         
        <div
          autocomlite="on"
          class="el-input"
        >
          <!---->
          <input
            autocomplete="off"
            class="el-input__inner"
            name="password"
            placeholder="密码"
            rows="2"
            type="password"
            validateevent="true"
          />
          <!---->
          <!---->
          <!---->
        </div>
         
        <span
          class="show-pwd"
        >
          <svg-icon
            svg-name="eye"
          />
        </span>
        <!---->
      </div>
    </div>
     
    <button
      class="el-button el-button--primary"
      style="width: 100%; margin-bottom: 30px;"
      type="button"
    >
      <!---->
      <!---->
      <span>
        登录
      </span>
    </button>
     
    <div
      class="tips"
    >
      账号:admin 密码随便填
    </div>
     
    <div
      class="tips"
    >
      账号:editer 密码随便填
    </div>
     
    <button
      class="el-button thirdparty-button el-button--primary"
      type="button"
    >
      <!---->
      <!---->
      <span>
        第三方登录
      </span>
    </button>
     
    <div
      class="el-dialog__wrapper"
      style="display: none;"
    >
      <div
        class="el-dialog"
      >
        <div
          class="el-dialog__header"
        >
          <span
            class="el-dialog__title"
          >
            第三方验证
          </span>
          <button
            aria-label="Close"
            class="el-dialog__headerbtn"
            type="button"
          >
            <i
              class="el-dialog__close el-icon el-icon-close"
            />
          </button>
        </div>
        <!---->
        <!---->
      </div>
    </div>
  </form>
</div>
`;
